va/validation-authority.go

```diff
@@ -288,14 +386,30 @@ func (va ValidationAuthorityImpl) validateDvsni(identifier core.AcmeIdentifier,
 	Z := hex.EncodeToString(h.Sum(nil))
 	ZName := fmt.Sprintf("%s.%s.%s", Z[:32], Z[32:], core.DVSNISuffix)
 
+	addr, allAddrs, problem := va.getAddr(identifier.Value)
+	challenge.ValidationRecord = &core.ValidationRecord{
+		Dvsni: &core.DvsniValidationRecord{
+			Hostname:          identifier.Value,
+			AddressesResolved: allAddrs,
+			AddressUsed:       addr,
+		},
+	}
+	if problem != nil {
+		challenge.Status = core.StatusInvalid
+		challenge.Error = problem
+		return challenge, challenge.Error
+	}
+
 	// Make a connection with SNI = nonceName
-	hostPort := identifier.Value + ":443"
+	hostPort := net.JoinHostPort(addr.String(), "443")
+	challenge.ValidationRecord.Dvsni.Port = "443"
 	if va.TestMode {
-		hostPort = "localhost:5001"
+		hostPort = net.JoinHostPort(addr.String(), "5001")
+		challenge.ValidationRecord.Dvsni.Port = "5001"
 	}
 	va.log.Notice(fmt.Sprintf("DVSNI [%s] Attempting to validate DVSNI for %s %s",
 		identifier, hostPort, ZName))
-	conn, err := tls.DialWithDialer(&net.Dialer{Timeout: 5 * time.Second}, "tcp", hostPort, &tls.Config{
+	conn, err := tls.DialWithDialer(&net.Dialer{Timeout: validationTimeout}, "tcp", hostPort, &tls.Config{
```

Comment:
`tls.DialWithDialer` doesn't take an interface but a `*net.Dialer` of which we cannot replace the `Dial` function. We could pull all the guts out of `DialWithDialer` to allow us to use our *sort of* dialer here, but then I think we are just adding more unnecessary code in balance.